src/GafferUI/GraphGadget.cpp

```diff
@@ -756,8 +1003,183 @@ ConnectionGadget *GraphGadget::reconnectionGadgetAt( const NodeGadget *gadget, c
 	return nullptr;
 }
 
+void GraphGadget::dirtyActive()
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+	m_activeStateDirty = true;
+	dirty( DirtyType::Render );
+}
+
+void GraphGadget::updateActive() const
+{
+	if( m_activeStateTask )
+	{
+		m_activeStateTask->cancelAndWait();
+		m_activeStateTask.reset();
+	}
+
+	tbb::spin_mutex::scoped_lock scopedLockA( m_pendingActiveStateMutex );
+	const Gaffer::Node *focusNode = m_scriptNode->getFocus();
+	std::vector<Gaffer::ConstPlugPtr> focusPlugs;
+	if( focusNode )
+	{
+		Gaffer::ConstPlugPtr hiddenFocusPlug = nullptr;
+		for( auto &plug : Gaffer::Plug::OutputRange( *focusNode ) )
+		{
+			const std::string &n = plug->getName();
+			if( n.size() >= 2 && n[0] == '_' && n[1] == '_' )
+			{
+				if( !hiddenFocusPlug )
+				{
+					hiddenFocusPlug = plug;
+				}
+				continue;
+			}
+
+			focusPlugs.push_back( plug );
+		}
+
+		if( !focusPlugs.size() && hiddenFocusPlug )
+		{
+			// If we couldn't find any visible output plug, take the hidden one
+			focusPlugs.push_back( hiddenFocusPlug );
+		}
+	}
+
+	if( !focusPlugs.size() )
+	{
+		m_pendingActiveState.reset();
+		scopedLockA.release();
+		Gaffer::ParallelAlgo::callOnUIThread(
+			[this] {
+				this->applyActive( false );
+			}
+		);
+		return;
+	}
+
+
+	tbb::spin_mutex &pendingActiveStateMutex = m_pendingActiveStateMutex;
+	std::unique_ptr<PendingActive> &pendingActiveState = m_pendingActiveState;
+
+	// Must hold a reference to stop us dying before our UI thread call is scheduled.
+	ConstGraphGadgetPtr thisRef = this;
+
+	m_activeStateTask = Gaffer::ParallelAlgo::callOnBackgroundThread(
+		// Subject:  TODO - does it matter that we're only including one plug as a subject?  We respond to any
+		// plugs being dirtied on the focus node with manual cancellation of this task, so hopefully it should
+		// be OK
+		focusPlugs[0].get(),
+		[focusPlugs, &pendingActiveStateMutex, &pendingActiveState, thisRef ] {
+			// We don't use the current context for our actual context - we use the script context
+			// and activePlugsAndNodes needs to continue even in cases where it can't figure out an accurate
+			// context andt just uses null ... but we do need to take the canceller from the context
+			// provided by callOnBackgroundThread
+			const IECore::Canceller *canceller = Gaffer::Context::current()->canceller();
+
+			Canceller::check( canceller );
+
+			std::unique_ptr<PendingActive> newActive( new PendingActive );
+			const Gaffer::ScriptNode *script = focusPlugs[0]->ancestor<Gaffer::ScriptNode>();
+			const Gaffer::Context *context = script ? script->context() : nullptr;
+
+			for( const Gaffer::ConstPlugPtr &focusPlug : focusPlugs )
+			{
+				activePlugsAndNodes(
+					focusPlug.get(), context, canceller,
+					newActive->activePlugs, newActive->activeNodes
+				);
+			}
+
+			{
+				tbb::spin_mutex::scoped_lock scopedLock( pendingActiveStateMutex );
+				pendingActiveState.swap( newActive );
```

Comment:
I've been trying to figure out why we need this `m_pendingActiveState` stuff, and I'm coming up blank. Can we just pass `newActive` directly as an argument to `applyActive()`?

Reply:
It's definitely true that after several iterations, this ended up overcomplicated.

I tried a naive implementation, and it works, but my debug output is showing that I'm actually evaluating it several times ... keeping track of the callOnUIThread being done by the background thread feels like a lot of things happening on different threads.  I guess with the current logic, the background thread has finishes, and launches a callOnUIThread, but then renderLayer comes along and notices that the active state is dirty, and that something should be computing it, but there is no background thread, because it's already finished, so it launches a new background thread.  I guess I can fix this with a new flag for "waiting on a callOnUIThread"?

Perhaps a better way of simplifying would be to keep pendingActiveState, but drop the callOnUIThread from the background thread, and instead, when the foreground thread hits renderLayer, it checks if a pending activation has been set before thinking about starting a background thread?  I think I must have originally been thinking of something like this, and I'm not sure why I didn't do it - maybe it felt wrong to stall in renderLayer while transferring the activate state to all the gadgets?

Reply:
It's not quite that simple - at the very least, we would need to catch the situation where m_backgroundTask still exists but is cancelled.  But maybe an approach like that could work, as long as we can absolutely depend on callOnUIThread being 100% reliable.

Reply:
Logic simplified quite a bit in 078a377